tgui/packages/common/color.ts

```diff
@@ -92,3 +92,478 @@ export class Color {
+    g: Number(match[3]) / (match[4] ? 100 / 255 : 1),
+    b: Number(match[5]) / (match[6] ? 100 / 255 : 1),
+    a: match[7] === undefined ? 1 : Number(match[7]) / (match[8] ? 100 : 1),
+  });
+};
+
+export const rgbStringToHsva = rgbaStringToHsva;
+
+const format = (number: number) => {
+  const hex = number.toString(16);
+  return hex.length < 2 ? '0' + hex : hex;
+};
+
+export const rgbaToHex = ({ r, g, b, a }: RgbaColor): string => {
+  const alphaHex = a < 1 ? format(round(a * 255)) : '';
+  return (
+    '#' + format(round(r)) + format(round(g)) + format(round(b)) + alphaHex
+  );
+};
+
+export const rgbaToHsva = ({ r, g, b, a }: RgbaColor): HsvaColor => {
+  const max = Math.max(r, g, b);
+  const delta = max - Math.min(r, g, b);
+
+  // prettier-ignore
+  const hh = delta
+    ? max === r
+      ? (g - b) / delta
+      : max === g
+        ? 2 + (b - r) / delta
+        : 4 + (r - g) / delta
+    : 0;
+
+  return {
+    h: 60 * (hh < 0 ? hh + 6 : hh),
+    s: max ? (delta / max) * 100 : 0,
+    v: (max / 255) * 100,
+    a,
+  };
+};
+
+export const roundHsva = (hsva: HsvaColor): HsvaColor => ({
+  h: round(hsva.h),
+  s: round(hsva.s),
+  v: round(hsva.v),
+  a: round(hsva.a, 2),
+});
+
+export const rgbaToRgb = ({ r, g, b }: RgbaColor): RgbColor => ({ r, g, b });
+
+export const hslaToHsl = ({ h, s, l }: HslaColor): HslColor => ({ h, s, l });
+
+export const hsvaToHsv = (hsva: HsvaColor): HsvColor => {
+  const { h, s, v } = roundHsva(hsva);
+  return { h, s, v };
+};
+
+const hexMatcher = /^#?([0-9A-F]{3,8})$/i;
+
+export const validHex = (value: string, alpha?: boolean): boolean => {
+  const match = hexMatcher.exec(value);
+  const length = match ? match[1].length : 0;
+
+  return (
+    length === 3 || // '#rgb' format
+    length === 6 || // '#rrggbb' format
+    (!!alpha && length === 4) || // '#rgba' format
+    (!!alpha && length === 8) // '#rrggbbaa' format
+  );
+};
+
+// Source for the following luminance and contrast calculation code: https://blog.cristiana.tech/calculating-color-contrast-in-typescript-using-web-content-accessibility-guidelines-wcag
+export const luminance = (rgb: RgbColor): number => {
+  const [r, g, b] = [rgb.r, rgb.g, rgb.b].map((v) => {
+    v /= 255;
+    return v <= 0.03928 ? v / 12.92 : Math.pow((v + 0.055) / 1.055, 2.4);
+  });
+  return r * 0.2126 + g * 0.7152 + b * 0.0722;
+};
+
+export const contrast = (
+  foreground: RgbColor,
+  background: RgbColor,
+): number => {
+  const foreground_luminance = luminance(foreground);
+  const background_luminance = luminance(background);
+  return background_luminance < foreground_luminance
+    ? (background_luminance + 0.05) / (foreground_luminance + 0.05)
+    : (foreground_luminance + 0.05) / (background_luminance + 0.05);
+};
+
+// Used for the colour picker, and anything else that wants a list of preset colours.
+export const colorList = [
+  ['003366', '336699', '3366CC', '003399', '000099', '0000CC', '000066'],
```

Comment:
Why is one list in one line and the rest are spread out? Keep it consistent

Reply:
Because there were too few elements in the line, VS Code automatically reformatted them into a single line upon saving. I didn't notice it, but I'll fix it.

Reply:
> Why is one list in one line and the rest are spread out? Keep it consistent

@Arturlang 

https://github.com/Bubberstation/Bubberstation/actions/runs/12399599657/job/34614884283

As I expected, the linters tests are failing when I try to provide smaller arrays in a different format, and the same happens with larger ones when I switch them to inline. For now, I’ve reverted everything back to the original setup. If you have any ideas on how to present this uniformly without triggering unit test errors, I’m open to suggestions.

Reply:
Formatting issues in TS files are 99.9% of the time a result of the sledgehammer effect prettier has. Add this comment on the line above this var so you can format it the way you want:
`// prettier-ignore`

Reply:
> Formatting issues in TS files are 99.9% of the time a result of the sledgehammer effect prettier has. Add this comment on the line above this var so you can format it the way you want: `// prettier-ignore`

Thanks for the tip! I tried it, and it works perfectly. I'll keep this in mind!